config_parser.cc

```diff
@@ -198,15 +199,21 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
       config_stack.top()->statements_.back().get()->child_block_.reset(
           new_config);
       config_stack.push(new_config);
+      num_open_blocks++;
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
+      if (last_token_type != TOKEN_TYPE_STATEMENT_END &&
+          last_token_type != TOKEN_TYPE_END_BLOCK &&
+          last_token_type != TOKEN_TYPE_START_BLOCK) {
         // Error.
         break;
       }
       config_stack.pop();
+      num_open_blocks--;
     } else if (token_type == TOKEN_TYPE_EOF) {
       if (last_token_type != TOKEN_TYPE_STATEMENT_END &&
-          last_token_type != TOKEN_TYPE_END_BLOCK) {
+          last_token_type != TOKEN_TYPE_END_BLOCK &&
```

Comment:
seems a little messsy to have a long AND statement and an OR statement at the end. If it doesn't hurt the logic, maybe you should try checking for num_open_blocks != 0 first ?